osquery/tables/system/linux/services.cpp

```diff
@@ -0,0 +1,774 @@
+      if (container_err == 0) {
+        break;
+      } else if (container_err < 0) {
+        throw std::runtime_error("Failed to enter the data container");
+      }
+
+      const char* property_name = nullptr;
+      if (sd_bus_message_read(reply, "s", &property_name) < 0) {
+        throw std::runtime_error("Failed to retrieve the property name");
+      }
+
+      if (std::strcmp(property_name, "MainPID") == 0) {
+        const char* contents;
+        if (sd_bus_message_peek_type(reply, NULL, &contents) < 0) {
+          throw std::runtime_error("Failed to determine the property type");
+        }
+
+        if (sd_bus_message_enter_container(
+                reply, SD_BUS_TYPE_VARIANT, contents) < 0) {
+          throw std::runtime_error("Failed to enter the variant container");
+        }
+
+        char property_type;
+        if (sd_bus_message_peek_type(reply, &property_type, NULL) < 0) {
+          throw std::runtime_error("Failed to determine the property type");
+        }
+
+        if (sd_bus_message_read_basic(reply, property_type, &process_id) < 0) {
+          throw std::runtime_error("Failed to read the property value");
+        }
+
+        if (sd_bus_message_exit_container(reply) < 0) {
+          throw std::runtime_error("Failed to exit the data container");
+        }
+
+      } else if (sd_bus_message_skip(reply, "v") < 0) {
+        throw std::runtime_error("Failed to skip the data entry");
+      }
+
+      if (sd_bus_message_exit_container(reply) < 0) {
+        throw std::runtime_error("Failed to exit the data container");
+      }
+    }
+
+    if (sd_bus_message_exit_container(reply) < 0) {
+      throw std::runtime_error("Failed to exit the data container");
+    }
+
+    reply = sd_bus_message_unref(reply);
+    return Status(0, "OK");
+
+  } catch (const std::exception& e) {
+    if (reply != nullptr) {
+      reply = sd_bus_message_unref(reply);
+    }
+
+    return Status(1, e.what());
+  }
+}
+
+struct SystemdUnitInfo final {
+  std::string id;
+  std::string description;
+  std::string load_state;
+  std::string active_state;
+  std::string sub_state;
+  std::string following;
+  std::string unit_path;
+  std::uint32_t job_id;
+  std::string job_type;
+  std::string job_path;
+  pid_t process_id;
+};
+
+Status getLoadedSystemdUnitList(std::vector<SystemdUnitInfo>& unit_list,
+                                sd_bus* bus) {
+  unit_list.clear();
+
+  sd_bus_message* message = nullptr;
+  sd_bus_message* reply = nullptr;
+  sd_bus_error bus_error = SD_BUS_ERROR_NULL;
+
+  try {
+    if (sd_bus_message_new_method_call(bus,
+                                       &message,
+                                       "org.freedesktop.systemd1",
+                                       "/org/freedesktop/systemd1",
+                                       "org.freedesktop.systemd1.Manager",
+                                       "ListUnitsFiltered")) {
```

Comment:
Do you need ListUnitsFiltered? In that case, ListUnits should be enough.

Reply:
I have tried to use ListUnits but it was causing issues on very old systemd versions; In the end I chose the most compatible route and used the same calls found on the command line tools (systemctl) installed in Ubuntu 15